quinn-udp/src/unix.rs

```diff
@@ -427,7 +480,8 @@ fn prepare_msg(
     }
 
     if let Some(ip) = &transmit.src_ip {
-        if cfg!(target_os = "linux") {
+        #[cfg(target_os = "linux")]
```

Comment:
No way to do this on FreeBSD?

Reply:
I have tried the similar operations on FreeBSD, but failed on most tests.
BTW, I'm not quite sure what this piece of code does.

Reply:
This sets the source IP address of an outgoing packet, which allows quinn to be well-behaved when an interface has multiple addresses on the same network. It's not required, but it's nice to have when possible.

Reply:
The IPv6 part works on FreeBSD with no change.
I tried to modify IPv4 part as below
```
                #[cfg(target_os = "freebsd")]
                IpAddr::V4(v4) => {
                    let addr = libc::in_addr {
                        s_addr: u32::from_ne_bytes(v4.octets()),
                    };
                    encoder.push(libc::IPPROTO_IP, libc::IP_SENDSRCADDR, addr);
                }
```
But some tests failed, please see the [log](https://github.com/quinn-rs/quinn/files/10088476/log.txt)

Reply:
Seems like that must not be the right way to do it, then. We can leave this for future work.

Reply:
I know what's wrong: sendmsg will fail with EINVAL if IP_SENDSRCADDR is used with a socket which is bound to something other than Ipv4Addr::UNSPECIFIED.
I added a check to make sure encode source ip in cmsg only when bind to Ipv4Addr::UNSPECIFIED.

Reply:
Hmm, what if the interface has multiple addresses bound to it? Does FreeBSD not support that?

Reply:
FreeBSD supports multiple IPs on a single interface.
But when a socket does not bind to Ipv4Addr::UNSPECIFIED (using bind()), FreeBSD's sendmsg just fails with EINVAL.
In the tests, eg. echo_v4, client binds to Ipv4Addr::UNSPECIFIED and server binds to Ipv4Addr::LOCALHOST, so client sendmsg works well but server fails with `sendmsg error: Os { code: 22, kind: InvalidInput, message: "Invalid argument" }`

Reply:
>  But when a socket does not bind to Ipv4Addr::UNSPECIFIED (using bind()), FreeBSD's sendmsg just fails with EINVAL.

Right, so on the surface this seems to prevent selecting the right IP address to respond from, if you have more than one address on the same subnet, which is weird. Not a blocker here, though, this is an improvement regardless.

Reply:
Nevermind, I confused myself, this behavior makes sense. Obviously if a socket is bound to a specific address you only need to send from that address.